### PR TITLE
Fixed: Unnecessary API Calls and Re-Renders on Patient Page

### LINKED ISSUE #9308: Unnecessary API Calls and Re-Renders on Patient Page
**Describe the bug**
Unnecessary API calls and re-rendering occur when entering the first digit into the PhoneNumber field. Additionally, the Patient page causes re-rendering when the search field is initially empty, and changing options also triggers a re-render.

**To Reproduce**

1. Go to the 'Facilities' page.
2. Then go to the 'Patient' page.
3. Open the network tab.
4. Put one digit inside the PhoneNumber field, change the option, or change the country code.
5. Notice in the network tab that an unnecessary API call is being made.

**Expected behavior**
For null values or when changing options, the API call should not be made. This is not expected behavior.

**Screenshots **
one digit input inside phonenuberfield
Image

while changing option
Image

while changiong country code
Image

**Desktop (please complete the following information):**

**Desktop (please complete the following information):**

- OS: windows
- Browser chrome
- Version - latest

**Additional context**
go to the facility page then come to patient page to reproduce the issue

### PR DESCRIPTION
**Pull request by i0am0arunava**
## Proposed Changes

- Fixes [#9308 ](https://github.com/ohcnetwork/care_fe/issues/9308)
prevent Unnecessary API Calls and Re-Renders on Patient Page

@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA

## Summary by CodeRabbit

- **New Features**
	- Introduced a new state variable to optimize API calls and enhance performance.

- **Bug Fixes**
	- Improved handling of duplicate patient IDs and refined state management.

- **Refactor**
	- Enhanced component structure for clarity and efficiency, reducing unnecessary re-renders.

### CODE REVIEW

File: src/components/Patient/ManagePatients.tsx [hunk 1]
```diff
@@ -291,8 +291,34 @@ export const PatientManager = () => {
     return cleanedData;
   };
 
+  // solving the issue Unnecessary API Calls and Re-Renders on Patient Page #9308
+  const [paramValue, setParamValue] = useState(params);
+  const previousValueRef = useRef(params);
+
+  function compareObjects(obj1: object, obj2: object) {
```
Comment (i0am0arunava) [inline src/components/Patient/ManagePatients.tsx:296]: If you console 'params' you observe that two params coming in differently but looking the same were being treated as different values. This was causing unnecessary API calls. Upon examining the previous parameter value and the current incoming parameter value, I found they were different. This issue was occurring not only when typing inside the phone number field but also when changing the option, causing unnecessary API calls.

```
{page: 1, limit: 12, name: undefined, patient_no: undefined, is_active: 'True', …}
{page: '1', limit: 12, name: undefined, patient_no: undefined, is_active: 'True', …}
```
If you observe the page key value, the previous one is a number, and the current one is a string. This was the problem. I solved it by comparing two objects while ignoring their value types, which now prevents unnecessary API calls

**Review by i0am0arunava — COMMENTED** [review]
fixed Unnecessary API Calls and Re-Renders on Patient Page

**Review by rithviknishad — CHANGES_REQUESTED** [review]
this is not solving the root cause. this is just hiding the root cause. 

the right solution here is to ensure `page` stays a number, just like the `limit` does.

**Comment by i0am0arunava** [comment]
> this is not solving the root cause. this is just hiding the root cause.
> 
> the right solution here is to ensure `page` stays a number, just like the `limit` does.

Got the point

**Comment by nihal467** [comment]
@i0am0arunava Can you please share the current status of the PR? Otherwise, the PR will be closed in the next 48 hours, and you will be unassigned from the issue due to inactivity.

**Comment by i0am0arunava** [comment]
> @i0am0arunava Can you please share the current status of the PR? Otherwise, the PR will be closed in the next 48 hours, and you will be unassigned from the issue due to inactivity.

@nihal467 @rithviknishad ,I am on leave due to final sem exam from  3rd to 17th December, and it was approved by Arvind . Please don't close this ,I will resume my work after 17th December

**Comment by rithviknishad** [comment]
Completed in #9400

**Pull request closed without merging**